File: bolao/shared/scripts/round_notification_ledger.mjs
/**
 * round_notification_ledger.mjs — ledger DURÁVEL de notificação POR RODADA (F6).
 *
 * ─── O QUE JÁ EXISTIA, E O QUE FALTAVA ───────────────────────────────────────────────────────
 *
 * A plataforma já tinha infraestrutura de notificação boa e consciente de segurança:
 * `notification_repository.mjs` (contrato + adaptadores memória/arquivo),
 * `durable_notification_repository.mjs` (Supabase via RPC, sem SELECT direto, sem coluna de
 * e-mail) e `bolao/shared/sql/010_notification_durability.sql` (claim atômico com
 * `for update skip locked`, lease, unicidade de idempotência, RLS sem policy para `anon`).
 *
 * Faltavam duas coisas, e as duas são o F6:
 *
 *   1. Aquilo tudo é POR DESTINATÁRIO. Uma "notificação de rodada" é uma unidade de negócio que
 *      abrange N destinatários e tem estado próprio — READY/PARTIAL/SENT dizem respeito à RODADA,
 *      não a um job isolado. Sem essa camada não existe resposta durável para "a rodada 22 foi
 *      comunicada?".
 *   2. Nenhum consumidor em produção. `grep` por essas classes em `bolao/br2026`, `bolao/cdb2026`,
 *      `bolao/copa2026` e `.github/workflows` retornava VAZIO. Era capacidade testada e nunca
 *      usada — o mesmo falso-verde arquitetural do FootballLiveStore.
 *
 * Este módulo é a camada 1. É PURO: recebe um repositório injetado e um relógio, não fala com
 * rede, não conhece Supabase e não conhece EmailJS.
 *
 * ─── IDENTIDADE ──────────────────────────────────────────────────────────────────────────────
 *
 *   rodada:       br2026:round-results:<round>:v1
 *   destinatário: br2026:round-results:<round>:v1#<entryRef>
 *
 * `entryRef` é o id OPACO da entrada. Nunca e-mail, nome, pagador ou transação — nem na chave,
 * nem no payload, nem no log.
 *
 * ─── GARANTIA DE ENTREGA (declarada, não presumida) ──────────────────────────────────────────
 *
 * O EmailJS não expõe consulta por idempotência: não dá para perguntar "você já aceitou a
 * mensagem X?". Portanto, no caso em que o provedor ACEITA e o processo morre antes de persistir,
 * não há como distinguir localmente "enviado" de "não enviado".
 *
 * A garantia honesta é, por destinatário:
 *
 *   - AT_MOST_ONCE até o provedor aceitar;
 *   - UNCERTAIN_AFTER_PROVIDER_ACCEPT na janela entre o aceite e a persistência.
 *
 * NÃO é exactly-once, e este módulo não finge que seja. A mitigação é conservadora: um job que
 * morreu com `SENDING` já registrado NÃO volta para a fila sozinho — vai para
 * `NEEDS_MANUAL_REVIEW`. Reenviar por conta própria porque a transação local se perdeu é
 * exatamente como se manda o mesmo e-mail duas vezes para onze pessoas.
 */

export const ROUND_STATE = Object.freeze({
  READY: "READY",
  CLAIMED: "CLAIMED",
  SENDING: "SENDING",
  PARTIAL: "PARTIAL",
  SENT: "SENT",
  FAILED: "FAILED",
  NEEDS_MANUAL_REVIEW: "NEEDS_MANUAL_REVIEW",
});

export const RECIPIENT_STATE = Object.freeze({
  PENDING: "PENDING",
  SENDING: "SENDING",
  ACCEPTED: "ACCEPTED",
  FAILED: "FAILED",
  BLOCKED: "BLOCKED",
  UNCERTAIN: "UNCERTAIN",
});

export const DELIVERY_SEMANTICS = "AT_MOST_ONCE_UNTIL_ACCEPT/UNCERTAIN_AFTER_PROVIDER_ACCEPT";

export function roundKey(round) {
  if (!Number.isInteger(round) || round < 1) throw new Error(`rodada invalida: ${round}`);
  return `br2026:round-results:${round}:v1`;
}

export function recipientKey(round, entryRef) {
  if (!entryRef || /@/.test(String(entryRef))) {
    // Uma chave com "@" quase certamente e um e-mail. A chave vai para log e para o banco.
    throw new Error("entryRef precisa ser um id opaco, nunca um endereco de e-mail");
  }
  return `${roundKey(round)}#${entryRef}`;
}

/** Hash determinístico e estável. Não é criptografia — é detecção de mudança. */
export function stableHash(value) {
  const text = typeof value === "string" ? value : JSON.stringify(value, Object.keys(value || {}).sort());
  let h1 = 0x811c9dc5, h2 = 0x01000193;
  for (let i = 0; i < text.length; i++) {
    const c = text.charCodeAt(i);
    h1 = Math.imul(h1 ^ c, 0x01000193) >>> 0;
    h2 = Math.imul(h2 + c, 0x85ebca6b) >>> 0;
  }
  return (h1.toString(16).padStart(8, "0") + h2.toString(16).padStart(8, "0"));
}

export function recipientSetHash(entryRefs) {
  return stableHash([...entryRefs].map(String).sort().join("|"));
}

export function fixtureSetHash(fixtureIds) {
  return stableHash([...fixtureIds].map(String).sort().join("|"));
}

const DEFAULT_LEASE_MS = 5 * 60 * 1000;

/**
 * Ledger de rodada sobre um repositório injetado.
 *
 * O repositório precisa de: get(key), put(key, record), listByPrefix(prefix), e
 * claimAtomic(key, owner, leaseUntil) devolvendo o registro reivindicado ou null.
 * `MemoryRoundLedgerRepo` abaixo implementa isso; a versão de produção usa as RPCs do
 * 010_notification_durability.sql, cuja atomicidade vem de `for update skip locked` no banco --
 * NUNCA de código sequencial aqui.
 */
export function createRoundLedger({ repo, now = () => Date.now(), leaseMs = DEFAULT_LEASE_MS, log = () => {} }) {
  function emit(eventType, record, extra) {
    // Só id opaco, hash e transição de estado. Nenhum e-mail, nome ou transação.
    log({
      eventType,
      idempotencyKey: record.idempotencyKey,
      roundNumber: record.roundNumber,
      state: record.state,
      expectedRecipientCount: record.expectedRecipientCount,
      resolvedRecipientCount: record.resolvedRecipientCount,
      contentHash: record.contentHash,
      recipientSetHash: record.recipientSetHash,
      attemptCount: record.attemptCount,
      ...extra,
    });
  }

  return {
    DELIVERY_SEMANTICS,

    /**
     * Cria o job da rodada se ainda não existir. Idempotente por construção.
     *
     * Se já existir com conteúdo ou conjunto de destinatários DIFERENTE, falha alto em vez de
     * mutar silenciosamente um trabalho em andamento — um e-mail já aceito não pode ser
     * retroativamente "sobre outra coisa".
     */
    async ensureJob({ roundNumber, expectedRecipientCount, entryRefs, contentHash, fixtureIds }) {
      const key = roundKey(roundNumber);
      const rsh = recipientSetHash(entryRefs);
      const fsh = fixtureSetHash(fixtureIds);
      const existing = await repo.get(key);

      if (existing) {
        const mudou = existing.contentHash !== contentHash
          || existing.recipientSetHash !== rsh
          || existing.fixtureSetHash !== fsh;
        if (mudou && existing.state !== ROUND_STATE.READY) {
          throw new Error(
            `CONTEUDO_MUDOU_EM_JOB_ATIVO: ${key} está em ${existing.state} com hashes diferentes. ` +
            `Superseda explicitamente (nova versão de chave) em vez de mutar um job em andamento.`
          );
        }
        if (mudou) {
          const atualizado = { ...existing, contentHash, recipientSetHash: rsh, fixtureSetHash: fsh,
                               expectedRecipientCount, updatedAt: now() };
          await repo.put(key, atualizado);
          emit("job_created", atualizado, { recreated: true });
          return atualizado;
        }
        return existing;
      }

      const record = {
        schemaVersion: 1,
        idempotencyKey: key,
        roundNumber,
        state: ROUND_STATE.READY,
        expectedRecipientCount,
        resolvedRecipientCount: entryRefs.length,
        contentHash, recipientSetHash: rsh, fixtureSetHash: fsh,
        recipients: entryRefs.map((ref) => ({ entryRef: ref, state: RECIPIENT_STATE.PENDING,
                                              providerMessageId: null, lastError: null })),
        attemptCount: 0,
        claimedBy: null, leaseUntil: null,
        createdAt: now(), updatedAt: now(), sentAt: null,
      };
      await repo.put(key, record);
      emit("job_created", record);
      return record;
    },

    /**
     * Portão de completude. Antes da PRIMEIRA chamada ao provedor:
     * esperados == resolvidos, senão zero envios e o ledger NÃO vira SENT.
     */
    async assertRecipientCompleteness(roundNumber) {
      const rec = await repo.get(roundKey(roundNumber));
      if (!rec) throw new Error(`job inexistente para a rodada ${roundNumber}`);
      if (rec.expectedRecipientCount !== rec.resolvedRecipientCount) {
        emit("job_blocked", rec, { reason: "RECIPIENT_SET_INCOMPLETE" });
        return { ok: false, reason: "RECIPIENT_SET_INCOMPLETE",
                 expected: rec.expectedRecipientCount, resolved: rec.resolvedRecipientCount };
      }
      return { ok: true };
    },

    /** Reivindicação atômica com lease. Exatamente um vencedor entre concorrentes. */
    async claim(roundNumber, owner) {
      const key = roundKey(roundNumber);
      const claimed = await repo.claimAtomic(key, owner, now() + leaseMs, now());
      if (!claimed) return null;
      emit("job_claimed", claimed, { owner });
      return claimed;
    },

    /** Marca início de envio. A partir daqui, uma morte de processo NÃO permite retry automático. */
    async markSending(roundNumber) {
      const key = roundKey(roundNumber);
      const rec = await repo.get(key);
      const next = { ...rec, state: ROUND_STATE.SENDING, attemptCount: rec.attemptCount + 1, updatedAt: now() };
      await repo.put(key, next);
      emit("job_send_started", next);
      return next;
    },

    async recordRecipient(roundNumber, entryRef, state, { providerMessageId = null, error = null } = {}) {
      const key = roundKey(roundNumber);
      const rec = await repo.get(key);
      const recipients = rec.recipients.map((r) =>
        r.entryRef === entryRef ? { ...r, state, providerMessageId, lastError: error } : r);
      const next = { ...rec, recipients, updatedAt: now() };
      await repo.put(key, next);
      emit(state === RECIPIENT_STATE.ACCEPTED ? "recipient_accepted" : "recipient_failed", next,
           { entryRef, recipientState: state });
      return next;
    },

    /** Deriva o estado da RODADA a partir das disposições por destinatário. */
    async settle(roundNumber) {
      const key = roundKey(roundNumber);
      const rec = await repo.get(key);
      const total = rec.recipients.length;
      const aceitos = rec.recipients.filter((r) => r.state === RECIPIENT_STATE.ACCEPTED).length;
      const incertos = rec.recipients.filter((r) => r.state === RECIPIENT_STATE.UNCERTAIN).length;

      let state;
      if (incertos > 0) state = ROUND_STATE.NEEDS_MANUAL_REVIEW;
      else if (aceitos === total && total > 0) state = ROUND_STATE.SENT;
      else if (aceitos > 0) state = ROUND_STATE.PARTIAL;
      else state = ROUND_STATE.FAILED;

      const next = { ...rec, state, updatedAt: now(),
                     sentAt: state === ROUND_STATE.SENT ? now() : rec.sentAt,
                     claimedBy: null, leaseUntil: null };
      await repo.put(key, next);
      emit(state === ROUND_STATE.SENT ? "job_sent"
           : state === ROUND_STATE.PARTIAL ? "job_partial" : "job_failed", next);
      return next;
    },

    /**
     * Recupera jobs cujo lease expirou (o runner morreu).
     *
     * REGRA CENTRAL DE SEGURANÇA: um job que morreu em SENDING não volta para a fila. Nessa
     * janela o provedor pode ter aceitado sem que a persistência acontecesse, e reenviar seria
     * duplicar. Vai para revisão manual, e os destinatários que estavam em SENDING viram
     * UNCERTAIN.
     */
    async recoverExpiredLeases() {
      const todos = await repo.listByPrefix("br2026:round-results:");
      const recuperados = [];
      for (const rec of todos) {
        if (!rec.leaseUntil || rec.leaseUntil > now()) continue;
        if (rec.state !== ROUND_STATE.CLAIMED && rec.state !== ROUND_STATE.SENDING) continue;

        const morreuEnviando = rec.state === ROUND_STATE.SENDING;
        const recipients = rec.recipients.map((r) =>
          morreuEnviando && r.state === RECIPIENT_STATE.SENDING
            ? { ...r, state: RECIPIENT_STATE.UNCERTAIN }
            : r);
        const next = {
          ...rec,
          recipients,
          state: morreuEnviando ? ROUND_STATE.NEEDS_MANUAL_REVIEW : ROUND_STATE.READY,
          claimedBy: null, leaseUntil: null, updatedAt: now(),
        };
        await repo.put(rec.idempotencyKey, next);
        emit("job_lease_expired", next, { diedWhileSending: morreuEnviando });
        recuperados.push(next);
      }
      return recuperados;
    },

    /** Rodadas ainda não comunicadas. Independentes entre si — nenhuma bloqueia outra. */
    async pendingRounds() {
      const todos = await repo.listByPrefix("br2026:round-results:");
      return todos
        .filter((r) => r.state !== ROUND_STATE.SENT && r.state !== ROUND_STATE.NEEDS_MANUAL_REVIEW)
        .sort((a, b) => a.roundNumber - b.roundNumber);
    },

    async get(roundNumber) {
      return repo.get(roundKey(roundNumber));
    },
  };
}

/** Repositório em memória. TESTE E DEV APENAS — produção usa as RPCs do Postgres. */
export function createMemoryRoundLedgerRepo() {
  const store = new Map();
  return {
    async get(key) { const v = store.get(key); return v ? JSON.parse(JSON.stringify(v)) : null; },
    async put(key, record) { store.set(key, JSON.parse(JSON.stringify(record))); },
    async listByPrefix(prefix) {
      return [...store.entries()].filter(([k]) => k.startsWith(prefix))
        .map(([, v]) => JSON.parse(JSON.stringify(v)));
    },
    /**
     * Reivindicação atômica. Em memória, "atômico" é trivial porque o Node é single-threaded —
     * e é justamente por isso que este adaptador NÃO prova concorrência. A prova real é o
     * `for update skip locked` da RPC no Postgres; ver o contrato estático em
     * test_round_notification_ledger.mjs.
     */
    async claimAtomic(key, owner, leaseUntil, nowMs) {
      const rec = store.get(key);
      if (!rec) return null;
      const leaseAtivo = rec.leaseUntil && rec.leaseUntil > nowMs;
      if (leaseAtivo) return null;
      if (rec.state !== "READY") return null;
      const next = { ...rec, state: "CLAIMED", claimedBy: owner, leaseUntil, updatedAt: nowMs };
      store.set(key, next);
      return JSON.parse(JSON.stringify(next));
    },
    _dump() { return [...store.values()]; },
  };
}

File: bolao/shared/scripts/round_notification_ledger.py
"""
round_notification_ledger.py — ledger durável de notificação POR RODADA (F6), lado Python.

Espelha `round_notification_ledger.mjs`. As duas implementações existem porque o caminho de
produção do e-mail de rodada é Python (`bolao/br2026/scripts/send_round_email.py`) enquanto os
gates e o worker são Node — é o mesmo padrão já adotado por `notification_repository.py`/`.mjs`.

DERIVA É O RISCO REAL AQUI, e este repositório já foi mordido por ela: em julho de 2026 uma
auditoria achou `send_result_email.py` com lógica de pontuação silenciosamente divergente do
`app.js`. Por isso `test_round_ledger_interop.mjs` executa AS DUAS implementações sobre os mesmos
casos e falha se um único hash, chave ou transição de estado divergir. Não é um teste de
formalidade: é a única coisa que impede este arquivo de virar uma segunda verdade.

Ver a docstring do .mjs para a justificativa da semântica de entrega — ela é a mesma, e a string
`DELIVERY_SEMANTICS` é comparada byte a byte pelo interop.
"""

import json

ROUND_STATE = {
    "READY": "READY",
    "CLAIMED": "CLAIMED",
    "SENDING": "SENDING",
    "PARTIAL": "PARTIAL",
    "SENT": "SENT",
    "FAILED": "FAILED",
    "NEEDS_MANUAL_REVIEW": "NEEDS_MANUAL_REVIEW",
}

RECIPIENT_STATE = {
    "PENDING": "PENDING",
    "SENDING": "SENDING",
    "ACCEPTED": "ACCEPTED",
    "FAILED": "FAILED",
    "BLOCKED": "BLOCKED",
    "UNCERTAIN": "UNCERTAIN",
}

DELIVERY_SEMANTICS = "AT_MOST_ONCE_UNTIL_ACCEPT/UNCERTAIN_AFTER_PROVIDER_ACCEPT"

_U32 = 0xFFFFFFFF


def round_key(round_number):
    if not isinstance(round_number, int) or isinstance(round_number, bool) or round_number < 1:
        raise ValueError(f"rodada invalida: {round_number}")
    return f"br2026:round-results:{round_number}:v1"


def recipient_key(round_number, entry_ref):
    if not entry_ref or "@" in str(entry_ref):
        raise ValueError("entryRef precisa ser um id opaco, nunca um endereco de e-mail")
    return f"{round_key(round_number)}#{entry_ref}"


def _imul(a, b):
    """Multiplicação de 32 bits com wraparound, como o `Math.imul` do JS.

    Sem isto o Python promoveria para inteiro arbitrário e os hashes divergiriam do lado Node —
    que é exatamente o tipo de deriva que o interop existe para pegar.
    """
    return ((a * b) & _U32)


def stable_hash(value):
    if isinstance(value, str):
        text = value
    else:
        text = json.dumps(value, sort_keys=True, separators=(",", ":"))
    h1, h2 = 0x811C9DC5, 0x01000193
    for ch in text:
        c = ord(ch)
        h1 = _imul(h1 ^ c, 0x01000193) & _U32
        h2 = _imul(h2 + c, 0x85EBCA6B) & _U32
    return f"{h1:08x}{h2:08x}"


def recipient_set_hash(entry_refs):
    return stable_hash("|".join(sorted(str(r) for r in entry_refs)))


def fixture_set_hash(fixture_ids):
    return stable_hash("|".join(sorted(str(f) for f in fixture_ids)))


DEFAULT_LEASE_MS = 5 * 60 * 1000


class MemoryRoundLedgerRepo:
    """TESTE E DEV APENAS. Produção usa as RPCs de `bolao/shared/sql/010_notification_durability.sql`,
    cuja atomicidade vem de `for update skip locked` no banco — nunca deste código."""

    def __init__(self):
        self._store = {}

    def get(self, key):
        v = self._store.get(key)
        return json.loads(json.dumps(v)) if v is not None else None

    def put(self, key, record):
        self._store[key] = json.loads(json.dumps(record))

    def list_by_prefix(self, prefix):
        return [json.loads(json.dumps(v)) for k, v in self._store.items() if k.startswith(prefix)]

    def claim_atomic(self, key, owner, lease_until, now_ms):
        rec = self._store.get(key)
        if rec is None:
            return None
        if rec.get("leaseUntil") and rec["leaseUntil"] > now_ms:
            return None
        if rec.get("state") != "READY":
            return None
        nxt = dict(rec, state="CLAIMED", claimedBy=owner, leaseUntil=lease_until, updatedAt=now_ms)
        self._store[key] = nxt
        return json.loads(json.dumps(nxt))


class RoundLedger:
    def __init__(self, repo, now=None, lease_ms=DEFAULT_LEASE_MS, log=None):
        self.repo = repo
        self._now = now or (lambda: 0)
        self.lease_ms = lease_ms
        self._log = log or (lambda e: None)

    DELIVERY_SEMANTICS = DELIVERY_SEMANTICS

    def _emit(self, event_type, record, **extra):
        # Só id opaco, hash e transição. Nenhum e-mail, nome, pagador ou transação.
        self._log(dict({
            "eventType": event_type,
            "idempotencyKey": record.get("idempotencyKey"),
            "roundNumber": record.get("roundNumber"),
            "state": record.get("state"),
            "expectedRecipientCount": record.get("expectedRecipientCount"),
            "resolvedRecipientCount": record.get("resolvedRecipientCount"),
            "contentHash": record.get("contentHash"),
            "recipientSetHash": record.get("recipientSetHash"),
            "attemptCount": record.get("attemptCount"),
        }, **extra))

    def ensure_job(self, round_number, expected_recipient_count, entry_refs, content_hash, fixture_ids):
        key = round_key(round_number)
        rsh = recipient_set_hash(entry_refs)
        fsh = fixture_set_hash(fixture_ids)
        existing = self.repo.get(key)

        if existing:
            mudou = (existing.get("contentHash") != content_hash
                     or existing.get("recipientSetHash") != rsh
                     or existing.get("fixtureSetHash") != fsh)
            if mudou and existing.get("state") != ROUND_STATE["READY"]:
                raise RuntimeError(
                    f"CONTEUDO_MUDOU_EM_JOB_ATIVO: {key} esta em {existing.get('state')} com hashes "
                    f"diferentes. Superseda explicitamente em vez de mutar um job em andamento."
                )
            if mudou:
                atualizado = dict(existing, contentHash=content_hash, recipientSetHash=rsh,
                                  fixtureSetHash=fsh, expectedRecipientCount=expected_recipient_count,
                                  updatedAt=self._now())
                self.repo.put(key, atualizado)
                self._emit("job_created", atualizado, recreated=True)
                return atualizado
            return existing

        record = {
            "schemaVersion": 1,
            "idempotencyKey": key,
            "roundNumber": round_number,
            "state": ROUND_STATE["READY"],
            "expectedRecipientCount": expected_recipient_count,
            "resolvedRecipientCount": len(entry_refs),
            "contentHash": content_hash,
            "recipientSetHash": rsh,
            "fixtureSetHash": fsh,
            "recipients": [{"entryRef": r, "state": RECIPIENT_STATE["PENDING"],
                            "providerMessageId": None, "lastError": None} for r in entry_refs],
            "attemptCount": 0,
            "claimedBy": None, "leaseUntil": None,
            "createdAt": self._now(), "updatedAt": self._now(), "sentAt": None,
        }
        self.repo.put(key, record)
        self._emit("job_created", record)
        return record

    def assert_recipient_completeness(self, round_number):
        rec = self.repo.get(round_key(round_number))
        if rec is None:
            raise RuntimeError(f"job inexistente para a rodada {round_number}")
        if rec["expectedRecipientCount"] != rec["resolvedRecipientCount"]:
            self._emit("job_blocked", rec, reason="RECIPIENT_SET_INCOMPLETE")
            return {"ok": False, "reason": "RECIPIENT_SET_INCOMPLETE",
                    "expected": rec["expectedRecipientCount"],
                    "resolved": rec["resolvedRecipientCount"]}
        return {"ok": True}

    def claim(self, round_number, owner):
        key = round_key(round_number)
        claimed = self.repo.claim_atomic(key, owner, self._now() + self.lease_ms, self._now())
        if not claimed:
            return None
        self._emit("job_claimed", claimed, owner=owner)
        return claimed

    def mark_sending(self, round_number):
        key = round_key(round_number)
        rec = self.repo.get(key)
        nxt = dict(rec, state=ROUND_STATE["SENDING"],
                   attemptCount=rec["attemptCount"] + 1, updatedAt=self._now())
        self.repo.put(key, nxt)
        self._emit("job_send_started", nxt)
        return nxt

    def record_recipient(self, round_number, entry_ref, state, provider_message_id=None, error=None):
        key = round_key(round_number)
        rec = self.repo.get(key)
        recipients = [dict(r, state=state, providerMessageId=provider_message_id, lastError=error)
                      if r["entryRef"] == entry_ref else r
                      for r in rec["recipients"]]
        nxt = dict(rec, recipients=recipients, updatedAt=self._now())
        self.repo.put(key, nxt)
        self._emit("recipient_accepted" if state == RECIPIENT_STATE["ACCEPTED"] else "recipient_failed",
                   nxt, entryRef=entry_ref, recipientState=state)
        return nxt

    def settle(self, round_number):
        key = round_key(round_number)
        rec = self.repo.get(key)
        total = len(rec["recipients"])
        aceitos = sum(1 for r in rec["recipients"] if r["state"] == RECIPIENT_STATE["ACCEPTED"])
        incertos = sum(1 for r in rec["recipients"] if r["state"] == RECIPIENT_STATE["UNCERTAIN"])

        if incertos > 0:
            state = ROUND_STATE["NEEDS_MANUAL_REVIEW"]
        elif aceitos == total and total > 0:
            state = ROUND_STATE["SENT"]
        elif aceitos > 0:
            state = ROUND_STATE["PARTIAL"]
        else:
            state = ROUND_STATE["FAILED"]

        nxt = dict(rec, state=state, updatedAt=self._now(),
                   sentAt=self._now() if state == ROUND_STATE["SENT"] else rec.get("sentAt"),
                   claimedBy=None, leaseUntil=None)
        self.repo.put(key, nxt)
        self._emit("job_sent" if state == ROUND_STATE["SENT"]
                   else "job_partial" if state == ROUND_STATE["PARTIAL"] else "job_failed", nxt)
        return nxt

    def recover_expired_leases(self):
        """Um job que morreu em SENDING NÃO volta para a fila — ver a docstring do .mjs."""
        recuperados = []
        for rec in self.repo.list_by_prefix("br2026:round-results:"):
            if not rec.get("leaseUntil") or rec["leaseUntil"] > self._now():
                continue
            if rec.get("state") not in (ROUND_STATE["CLAIMED"], ROUND_STATE["SENDING"]):
                continue
            morreu_enviando = rec["state"] == ROUND_STATE["SENDING"]
            recipients = [dict(r, state=RECIPIENT_STATE["UNCERTAIN"])
                          if morreu_enviando and r["state"] == RECIPIENT_STATE["SENDING"] else r
                          for r in rec["recipients"]]
            nxt = dict(rec, recipients=recipients,
                       state=ROUND_STATE["NEEDS_MANUAL_REVIEW"] if morreu_enviando else ROUND_STATE["READY"],
                       claimedBy=None, leaseUntil=None, updatedAt=self._now())
            self.repo.put(rec["idempotencyKey"], nxt)
            self._emit("job_lease_expired", nxt, diedWhileSending=morreu_enviando)
            recuperados.append(nxt)
        return recuperados

    def pending_rounds(self):
        todos = self.repo.list_by_prefix("br2026:round-results:")
        pend = [r for r in todos
                if r["state"] not in (ROUND_STATE["SENT"], ROUND_STATE["NEEDS_MANUAL_REVIEW"])]
        return sorted(pend, key=lambda r: r["roundNumber"])

    def get(self, round_number):
        return self.repo.get(round_key(round_number))
